ic-asset/src/asset_config.rs

```diff
@@ -0,0 +1,627 @@
+use anyhow::bail;
+use derivative::Derivative;
+use globset::{Glob, GlobMatcher};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::{
+    cell::{RefCell, RefMut},
+    collections::HashMap,
+    fs::File,
+    io::BufReader,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+const ASSETS_CONFIG_FILENAME: &str = ".ic-assets.json";
+
+thread_local! {
+    // There is no simple way to pass state into serde deserializer. Using this as a walkaround.
+    // https://stackoverflow.com/questions/54052495/mock-instance-inside-serde-implementation
+    static CURRENTLY_PROCESSED_ASSETS_DIRECTORY: RefCell<PathBuf> = RefCell::new(PathBuf::new());
+}
+
+pub(crate) type HeadersConfig = HashMap<String, Value>;
+type Map = HashMap<PathBuf, Arc<AssetConfigTreeNode>>;
+
+#[derive(Deserialize, Serialize, Debug, Default, Clone, PartialEq, Eq)]
+pub(crate) struct CacheConfig {
+    pub(crate) max_age: u64,
+}
+
+#[derive(Deserialize, Clone, Derivative)]
+#[derivative(Debug)]
+struct AssetConfigRule {
+    #[serde(deserialize_with = "string_to_glob")]
+    #[derivative(Debug(format_with = "fmt_glob_field"))]
+    r#match: GlobMatcher,
+    cache: Option<CacheConfig>,
+    #[serde(default, deserialize_with = "deser_headers")]
+    headers: Maybe<HeadersConfig>,
+}
+
+#[derive(Deserialize, Clone, Debug)]
+enum Maybe<T> {
+    Null,
+    Absent,
```

Comment:
Such a setup allows for clearing the headers. 
- `null` clears the headers
- `no field` doesn't affect the outcome
- `"headers": {}` doesn't affect the outcome

For example, when having:
```
/.ic-assets.json
/file.html
/abc/.ic-assets.json
/abc/file.html
```
and the following rule in the top-level config file:
```
{
  "match": "*",
  "headers": {
    "something": "abc"
  }
}
```

The rule would apply to `/abc/file.html`, whereas it may not be desirable. In order to clear the headers, the user could define the following rule in e.g. `/abc/.ic-assets.json`:
```
{
  "match": "file.html",
  "headers": null
}
```

Reply:
made it clearer & self-documented no longer using type casting `from` `into` and instead did this:
```
impl AssetConfig {
    fn merge(mut self, other: &AssetConfigRule) -> Self {
        if let Some(c) = &other.cache {
            self.cache = Some(c.to_owned());
        };
        match (self.headers.as_mut(), &other.headers) {
            (Some(sh), Maybe::Value(oh)) => sh.extend(oh.to_owned()),
            (None, Maybe::Value(oh)) => self.headers = Some(oh.to_owned()),
            (_, Maybe::Null) => self.headers = None,
            (_, Maybe::Absent) => (),
        };
        self
    }
}
```

Reply:
Can you describe these distinctions in the changelog entry?
```
null clears the headers
no field doesn't affect the outcome
"headers": {} doesn't affect the outcome
```
It might also be good to describe the rationale for the difference between `null` and `{}` handling.  The rationale isn't clear to me.

Reply:
should both `null` and `{}` behave similarly?

Reply:
I'm not married to the distinction, but I'd like to see a way to "undo" the headers

Reply:
I would suggest so, yes, or a separate field to indicate that all prior headers should be cleared.